[IMP] stock_quant_package_product_packaging: set package type

gurneyalex opened the pull request:
This module includes code extracted from `stock_storage_type` (https://github.com/OCA/stock-logistics-putaway/) because it can be useful to get a value for the `package_type_id` field of `stock.quant.package` without pulling the storage type logic.

gurneyalex committed:
[IMP] stock_quant_package_product_packaging

Extract code from `stock_storage_type`
(https://github.com/OCA/stock-logistics-putaway/) because it can be
useful to get a value for the `package_type_id` field of
`stock.quant.package` without pulling the storage type logic.

[IMP] better assignment of the package type

Use the quantity in the stock quant package to find the best suitable
product.packaging and set the package_type from there.

fixes https://github.com/OCA/stock-logistics-workflow/issues/967

jbaudoux commented on stock_quant_package_package_type/models/stock_quant_package.py:
```diff
@@ -0,0 +1,52 @@
+                # to not trigger constraint like height requirement
+                # (we are delivering them, not storing them)
+                continue
+            package_type = package.product_packaging_id.package_type_id
```
Here you need to enhance when package doesn't have a product_packaging_id set because the quantity is a multiple of a packaging quantity. The package type of the corresponding product packaging multiple should be set. Relates to that issue https://github.com/OCA/stock-logistics-workflow/issues/967
and to this line https://github.com/OCA/stock-logistics-tracking/blob/18.0/stock_quant_package_product_packaging/models/stock_quant_package.py#L49
You need to do a search with 
`("qty", "<", pack.single_product_qty),` order by qty desc sequence desc, limit 1
if qty is a multiple of pack.single_product_qty, then you found a packaging and you can take its package type.

The second issue is that the package type needs to come earlier in the process. `product_packaging_id` is set based on the quants of the package, so on action_done on a move line.
As the purpose of the package type is to compute a putaway, we need the value to be set or get updated when the package is set as destination package of a move line.

gurneyalex replied:
@jbaudoux see https://github.com/OCA/stock-logistics-tracking/pull/49/commits/24e5889836536550756a91a3787987ef5c3e0cc7

grindtildeath commented on stock_quant_package_package_type/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+{
+    "name": "Stock Quant Package Package Type",
+    "summary": "Use dimensions on packages",
+    "version": "18.0.1.0.1",
```
```suggestion
    "version": "18.0.1.0.0",
```

grindtildeath commented on stock_quant_package_package_type/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+{
+    "name": "Stock Quant Package Package Type",
+    "summary": "Use dimensions on packages",
```
```suggestion
    "summary": "Use package type from packaging on packages",
```

grindtildeath reviewed (approved):
LG, but I'm not sure if module name couldn't be improved.

On one hand I find the name not very explicit because there's already a `package_type_id` field on `stock.quant.package` defined in the stock module, and on the other hand I don't think `stock_quant_package_packaging_package_type` makes it any easier to understand :shrug: 

Pre approving

jbaudoux commented on stock_quant_package_package_type/models/product_product.py:
```diff
@@ -0,0 +1,18 @@
+    def _find_best_packaging(self, quantity):
+        self.ensure_one()
+        packagings = self.env["product.packaging"].search(
+            [("product_id", "=", self.id)], order="qty DESC"
```
Just because we can have multiple packaging for the same qty
```suggestion
            [("product_id", "=", self.id)], order="qty DESC, sequence ASC"
```

TDu replied:
```suggestion
            [("product_id", "=", self.id), ("qty", "<=", quantity)], order="qty DESC"
```

jbaudoux commented on stock_quant_package_package_type/models/stock_quant_package.py:
```diff
@@ -0,0 +1,55 @@
+                # to not trigger constraint like height requirement
+                # (we are delivering them, not storing them)
+                continue
+            package_type = package.single_product_id.package_type_id
```
I'm not convinced by this fallback on the package type of the product. We don't fallback on smaller packagings either.
Anything broken if you remove this?

gurneyalex replied:
"anything" in which scope? I have no idea... 

In the scope of the original module, not having this breaks the `test_autoassign_no_packaging` test, which you flag as "to remove". But I don't know if projects are depending on this feature.

gurneyalex replied:
In the scope of the original module, not having this breaks the test_autoassign_no_packaging () test

jbaudoux replied:
Test stock storage type from which you extracted the code.
I'm in favor of removing it, maybe in separate commit then.

gurneyalex replied:
I extracted the test from that module...

jbaudoux commented on stock_quant_package_package_type/models/product_template.py:
```diff
@@ -0,0 +1,15 @@
+from odoo import fields, models
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    package_type_id = fields.Many2one(
+        "stock.package.type",
+        string="Package type",
+        help="Defines a 'default' package type for this product to be "
+        "applied on packages without product packagings and on put-away "
+        "computation based on package type for product not in a package",
+    )
```
Then this could be removed from this module. It's strange to have it here

jbaudoux commented on stock_quant_package_package_type/tests/test_auto_assign_package_type.py:
```diff
@@ -0,0 +1,133 @@
+        )
+        self.assertEqual(package.package_type_id, self.package_type)
+
+    def test_auto_assign_packaging(self):
```
this test would be obsolete

jbaudoux commented on stock_quant_package_package_type/tests/test_auto_assign_package_type.py:
```diff
@@ -0,0 +1,133 @@
+
+        self.assertEqual(self.auto_assigned_package_type, first_package.package_type_id)
+
+    def test_auto_assign_no_packaging(self):
```
and this one also

jbaudoux reviewed (commented):
Do we break something in stock_storage_type with such change?

TDu commented on stock_quant_package_package_type/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "depends": [
+        "stock_quant_package_product_packaging",
+    ],
+    "data": [],
```
```suggestion
```

TDu commented on stock_quant_package_package_type/models/product_product.py:
```diff
@@ -0,0 +1,18 @@
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    def _find_best_packaging(self, quantity):
```
IMO, this function would make more sense in `stock_quant_package_product_packaging` the same computation is needed there as well https://github.com/OCA/stock-logistics-tracking/blob/6a61dd328d1cb61730934e15bd5805cc5d85f8da/stock_quant_package_product_packaging/models/stock_quant_package.py#L46, no ?

jbaudoux commented on stock_quant_package_package_type/models/stock_quant_package.py:
```diff
@@ -0,0 +1,55 @@
+                continue
+            package.package_type_id = package_type
+
+    def _sync_package_type_from_single_product(self):
```
I would pass the product and quantity to this method. Then it can be called before package.single_product_id is set, i.e. before the move is done. Will be required to compute the put-away

jbaudoux reviewed (commented):
LG. You cleanup the commits? Add the module name as prefix in the commit messages

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -52,3 +64,35 @@ def auto_assign_packaging(self):
+                continue
+            package.package_type_id = package_type
+
+    def _sync_package_type_from_single_product(self):
```
@TDu will still need to call this method and give a product in argument as `package.single_product_id` won't be yet computed

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -52,3 +64,35 @@ def auto_assign_packaging(self):
+
+    def _sync_package_type_from_single_product(self):
+        for package in self:
+            if package.package_type_id:
+                # Do not set package type for delivery packages
+                # to not trigger constraint like height requirement
+                # (we are delivering them, not storing them)
+                continue
```
When the package gets empties (quants are completely removed), the package_type_id must be removed otherwise you cannot recycle the package for another product with a different package type

gurneyalex replied:
done, by overriding `stock.quant` method `move_quants` (I'll check your comment on that diff chunk)

gurneyalex committed:
[IMP] stock_quant_package_product_packaging: fix the tests and handle unpacking

when a package is unpacked, the package type must be reset

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant.py:
```diff
@@ -0,0 +1,16 @@
+class StockQuant(models.Model):
+    _inherit = "stock.quant"
+
+    def move_quants(
+        self, location_dest_id=False, package_dest_id=False, message=False, unpack=False
+    ):
+        package_ids = self.package_id
+        res = super().move_quants(location_dest_id, package_dest_id, message, unpack)
+        if unpack:
+            package_ids.package_type_id = False
+        return res
```
For emptying the package type, you likely also need to hook _action_done on stock.move.line. Take a backup of the source packages, call super, check if there are no more quants != 0
Here also you need to check if the package is empty. I think you can move some of the package quants with unpack=True. So you still have some quants in the package.

gurneyalex replied:
thanks :pray:

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
+    def _assign_package_type(self, product, quantity):
+        self.ensure_one()
+        packaging = product._find_best_packaging(quantity)
+        if packaging and packaging.qty == quantity:
+            self.write({"product_packaging_id": packaging.id})
```
add a note that `_sync_package_type_from_packaging`will be called by `write`

jbaudoux replied:
```suggestion
        if packaging and packaging.qty == quantity:
            self.product_packaging_id = packaging
        elif self.product_packaging_id:
            self.product_packaging_id = False
```

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
-                if packaging:
-                    pack.write({"product_packaging_id": packaging.id})
+
+    def _assign_package_type(self, product, quantity):
```
@TDu This is the method you need to call to set the package type on the move line result package

gurneyalex replied:
@TDu the method was just renamed to `_assign_packaging` :warning:

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
-                        ("qty", "=", pack.single_product_qty),
-                    ],
-                    limit=1,
+                pack._assign_package_type(
```
```suggestion
                pack._assign_packaging(
```

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
-                if packaging:
-                    pack.write({"product_packaging_id": packaging.id})
+
+    def _assign_package_type(self, product, quantity):
```
```suggestion
    def _assign_packaging(self, product, quantity):
```

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
+        elif packaging:
+            self._sync_package_type_from_packaging(packaging)
+        else:
+            self._sync_package_type_from_single_product()
```
```suggestion
            self._sync_package_type_from_single_product(product, quantity)
```

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
+                continue
+            package.package_type_id = package_type
+
+    def _sync_package_type_from_single_product(self):
```
```suggestion
    def _sync_package_type_from_single_product(self, product, quantity):
```

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
+
+    def _sync_package_type_from_single_product(self):
+        for package in self:
+            if package.single_product_id and package.package_type_id:
```
there is a product by definition
```suggestion
            if package.package_type_id:
```

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
+                continue
+
+            package_type = package.single_product_id.package_type_id
+            package.package_type_id = package_type
```
```suggestion
            package_type = product.package_type_id
            if package_type:
                package.package_type_id = package_type
```

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
                 )
-                if packaging:
-                    pack.write({"product_packaging_id": packaging.id})
+
```
I would remove `not pack.product_packaging_id` from the `if`
and add 
```            
              else:
                   self.product_packaging_id = False
```
This is because this method is called on action done. So the first product you put in the pack will trigger this method and may set a packaging. The next product you put in the pack, you should not keep the packaging

gurneyalex replied:
I guess this was an explicit suggestion inside the `_assign_packaging` method, so I think this is done.

jbaudoux replied:
@gurneyalex single_product_id/qty are computed fields but product_packaging_id not. So when you add a product to a package, single_product_id will become false. I don't see where product_packaging_id gets to False.

jbaudoux replied:
I added a test for this and pushed a fix on this PR

jbaudoux commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,42 @@ def auto_assign_packaging(self):
+        packaging = product._find_best_packaging(quantity)
+        if packaging and packaging.qty == quantity:
+            self.write({"product_packaging_id": packaging.id})
+        elif packaging:
```
```suggestion
        if packaging:
```

jbaudoux reviewed (commented):
Update product_packaging_id when package gets more products

gurneyalex committed:
[REF] stock_quant_package_product_packaging

rename method _assign_packaging_type to _assign_packaging

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

jbaudoux committed:
[REF] stock_quant_package_product_packaging: review comments

Allow to call package type computation before action done.
Required for put-away computation.

gurneyalex committed:
[FIX] stock_quant_package_product_packaging: package type

missing case of empty package checking

ivantodorovich commented:
A test for `_sync_package_type_from_single_product` would be nice to have, since that branch is missing

TDu commented:
I see some errors https://github.com/OCA/stock-logistics-shopfloor/pull/40#pullrequestreview-3165586958 when using this PR in https://github.com/OCA/stock-logistics-shopfloor/pull/40, not sure why ?

TDu commented on stock_quant_package_product_packaging/models/stock_quant_package.py:
```diff
@@ -43,12 +56,46 @@ def auto_assign_packaging(self):
+        if not self.package_type_id and product.package_type_id:
+            self.package_type_id = product.package_type_id
+
+    def _sync_package_type_from_packaging(self, packaging):
```
This declaration clashes with this one https://github.com/OCA/stock-logistics-putaway/blob/1043b255ccc94cc137d8d3e4360601daae53527b/stock_storage_type/models/stock_quant_package.py#L70
(same function name, different parameter list) If that other module is installed

jbaudoux replied:
@TDu you need to take the PR where it was removed from storage type

jbaudoux replied:
https://github.com/OCA/stock-logistics-putaway/pull/11/files

TDu replied:
oh..ok

simahawk commented:
@jbaudoux ping

jbaudoux committed:
[FIX] stock_quant_package_product_packaging: update product_packaging_id

jbaudoux commented:
/ocabot merge minor
